pinot-core/src/main/java/org/apache/pinot/core/query/scheduler/SecondaryWorkloadQueue.java

```diff
@@ -0,0 +1,167 @@
+
+  private static int _wakeUpTimeMs;
+  private final int _maxPendingPerGroup;
+
+  private final SchedulerGroup _schedulerGroup;
+
+  private final Lock _queueLock = new ReentrantLock();
+  private final Condition _queryReaderCondition = _queueLock.newCondition();
+  private final ResourceManager _resourceManager;
+  private final int _queryDeadlineMs;
+
+  public SecondaryWorkloadQueue(PinotConfiguration config, ResourceManager resourceManager) {
+    Preconditions.checkNotNull(config);
+    Preconditions.checkNotNull(resourceManager);
+
+    _queryDeadlineMs =
+        config.getProperty(SECONDARY_QUEUE_QUERY_TIMEOUT, DEFAULT_SECONDARY_QUEUE_QUERY_TIMEOUT_SEC) * 1000;
+    _wakeUpTimeMs = config.getProperty(QUEUE_WAKEUP_MS, DEFAULT_WAKEUP_MS);
+    _maxPendingPerGroup = config.getProperty(MAX_PENDING_SECONDARY_QUERIES, DEFAULT_MAX_PENDING_SECONDARY_QUERIES);
+    LOGGER.info("queryDeadlineMs={}, wakeupTimeMs={},maxPendingPerGroup={}", _queryDeadlineMs, _wakeUpTimeMs,
+        _maxPendingPerGroup);
+    _schedulerGroup = new FCFSSchedulerGroup(SECONDARY_WORKLOAD_GROUP_NAME);
+    _resourceManager = resourceManager;
+  }
+
+  /**
+   * Adds a query to the secondary workload queue.
+   * @param query
+   * @throws OutOfCapacityException
+   */
+  public void put(SchedulerQueryContext query)
+      throws OutOfCapacityException {
+    Preconditions.checkNotNull(query);
+    _queueLock.lock();
+    try {
+      checkSchedulerGroupCapacity(query);
+      query.setSchedulerGroupContext(_schedulerGroup);
+      _schedulerGroup.addLast(query);
+      _queryReaderCondition.signal();
+    } finally {
+      _queueLock.unlock();
+    }
+  }
+
+  /**
+   * Blocking call to read the next query
+   * @return
+   */
+  @Nullable
+  public SchedulerQueryContext take() {
+    _queueLock.lock();
+    try {
+      while (true) {
+        SchedulerQueryContext schedulerQueryContext;
+        while ((schedulerQueryContext = takeNextInternal()) == null) {
+          try {
+            _queryReaderCondition.await(_wakeUpTimeMs, TimeUnit.MILLISECONDS);
+          } catch (InterruptedException e) {
+            return null;
+          }
+        }
+        return schedulerQueryContext;
+      }
+    } finally {
+      _queueLock.unlock();
+    }
+  }
+
+  public List<SchedulerQueryContext> drain() {
+    List<SchedulerQueryContext> pending = new ArrayList<>();
+    _queueLock.lock();
+    try {
+      while (!_schedulerGroup.isEmpty()) {
+        pending.add(_schedulerGroup.removeFirst());
+      }
+    } finally {
+      _queueLock.unlock();
+    }
+    return pending;
+  }
+
+  private SchedulerQueryContext takeNextInternal() {
+    long startTimeMs = System.currentTimeMillis();
+    long deadlineEpochMillis = startTimeMs - _queryDeadlineMs;
+
+    _schedulerGroup.trimExpired(deadlineEpochMillis);
+    if (_schedulerGroup.isEmpty() || !_resourceManager.canSchedule(_schedulerGroup)) {
+      return null;
+    }
+
+    if (LOGGER.isDebugEnabled()) {
```

Comment:
Is this block really needed ?

I don't think there is a need to check for if condition. Based on the level, it should automatically know when to log or not.

Reply:
The "if" block is added so that we don't pay the overhead of the computation inside the if block before the DEBUG log.